### DISCUSSION TITLE
A dispatcher that can chain call dataLoaders in one fetcher. Hope it can help everyone in need.

### DISCUSSION THREAD

**Discussion by blabla-yy** [graphql-java/graphql-java#3362] [General]
Hi
I have been using GraphQL-Java for a long time, and during my use, the fetcher may need to trigger multiple dataloaders in a chain. I want to share a dispatcher I implemented based on Node.js next-tick.
https://github.com/blabla-yy/graphql-java-dispatcher

Features:
1. Supports asynchronous and chained calls to DataLoader in GraphQL-Java Fetcher.
2. Single-threaded asynchronous, more suitable for IO-heavy projects. You can have all fetchers run in one thread during a GraphQL execution, thus avoiding multi-threading issues. ThreadLocal, etc. are also supported.
3. Does not rely on instrumentation, has no counting and concurrent locks, so the performance is better.
4. It will not cause dataloader future "never complete".

Since the user is required to consciously control the Executor of each CompletableFuture in Fetcher, it is somewhat intrusive. May not be suitable for all users. I provide a simple wrapper class that can reduce the perception of Executor usage. 

Hope it can help everyone in need

**Comment by bbakerman** [comment]
Hmm this is very interesting.  I didn't look into the base class of this `io.github.blabla-yy:current-thread-executor` but being based on a `nextTick()` like node is an interesting idea.

For the record the latest version of java dataloader has a `Executor` based mechanism that allows chained data loader calls.  

It does this by submitting calls to "dispatch" continuously.  This is the opposite to what you have done above.  In a single thread tick based code, all dataloader calls MUST be submitted as jobs into a central message queue.

The work above takes the opposite approach and takes a polling approach where the ScheduledDataLoaderRegistry is asked **periodically AND continuously** if it wants to dispatch via a `DispatchPredicate`


https://github.com/graphql-java/java-dataloader#chaining-dataloader-calls

**Reply by blabla-yy** [reply]
Sorry, I didn't realize that the new version of dataLoader already supports this feature.
I read the usage and implementation, and it supports richer usage scenarios. I have also seen the implementation of Golang before, which also uses a similar time window solution.

My implementation is relatively simple. Since the Executor is customized, dispatchAll() can be executed after all tasks are completed, and then wait and execute a new round of tasks to arrive. Repeat until GraphQL execution is complete.
Therefore, a custom Executor must be specified, but the requests are aggregated as much as possible without any additional overhead.
